allennlp/interpret/attackers/hotflip.py

```diff
@@ -0,0 +1,203 @@
+            indexed_tokens = indexer.tokens_to_indices(tokens, self.vocab, "token_characters")
+            padded_tokens = indexer.pad_token_sequence(indexed_tokens,
+                                                       desired_num_tokens={"token_characters": len(tokens)},
+                                                       padding_lengths={"num_token_characters": max_token_length})
+            all_inputs['token_characters'] = torch.LongTensor(padded_tokens['token_characters']).unsqueeze(0)
+
+            if "elmo" in self.predictor._dataset_reader._token_indexers:
+                elmo_tokens = []
+                indexer = self.predictor._dataset_reader._token_indexers["elmo"]
+                for tok in all_tokens:
+                    tmp = indexer.tokens_to_indices([Token(text=tok)], self.vocab, "sentence")["sentence"]
+                    elmo_tokens.append(tmp[0])
+                all_inputs["elmo"] = torch.LongTensor(elmo_tokens).unsqueeze(0)
+
+        # find the TextFieldEmbedder
+        for module in self.predictor._model.modules():
+            if isinstance(module, TextFieldEmbedder):
+                embedder = module
+        # pass all tokens through the fake matrix and create an embedding out of it.
+        embedding_matrix = embedder(all_inputs).squeeze()
+        return Embedding(num_embeddings=self.vocab.get_vocab_size('tokens'),
+                         embedding_dim=embedding_matrix.shape[1],
+                         weight=embedding_matrix,
+                         trainable=False)
+
+    def attack_from_json(self,
+                         inputs: JsonDict = None,
+                         name_of_input_field_to_attack: str = 'tokens',
+                         name_of_grad_input_field: str = 'grad_input_1',
+                         ignore_tokens: List[str] = ["@@NULL@@"]): # pylint disable=unused-argument,dangerous-default-value
+
+        """
+        Replaces one token at a time from the input until the model's prediction changes.
+        `name_of_input_field_to_attack` is for example `tokens`, it says what the input
+        field is called. `name_of_grad_input_field` is for example `grad_input_1`, which
+        is a key into a grads dictionary.
+
+        The method computes the gradient w.r.t. the tokens, finds
+        the token with the maximum gradient (by L2 norm), and replaces it will
+        another token based on the first-order Taylor approximation of the loss.
+        This process is iteratively repeated until the prediction changes.
+        Once a token is replaced, it is not flipped again.
+
+        TODO (@Eric-Wallace) add functionality for ignore_tokens in the future.
+        """
+        original_instances = self.predictor.inputs_to_labeled_instances(inputs)
+        original_tokens = list(original_instances[0][name_of_input_field_to_attack].tokens)
+        final_tokens = []
+        new_prediction = None
+        for new_instances in original_instances:
```

Comment:
`new_instance` here, then `new_instances` when you put it into a list?  Actually, why do you put it into a list?  It looks like you always query it as `new_instances[0]`.  Just call `get_gradients([new_instance])` - that looks like the only time you actually use it as a list.